pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1507,24 +1556,42 @@ public CompletableFuture<Void> close(
         CompletableFuture<Void> closeFuture = new CompletableFuture<>();
 
         lock.writeLock().lock();
+        CompletableFuture<Void> inProgressTransferCloseTask = null;
         try {
             if (!disconnectClients) {
                 transferring = true;
             }
             // closing managed-ledger waits until all producers/consumers/replicators get closed. Sometimes, broker
             // forcefully wants to close managed-ledger without waiting all resources to be closed.
-            if (!isClosingOrDeleting || closeWithoutWaitingClientDisconnect) {
-                fenceTopicToCloseOrDelete();
+            if (isClosingOrDeleting) {
+                if (!disconnectClients) {
+                   return closeFutures.notDisconnectClients;
+                }
+                /** There is a in-progress half closing task. see the section 2-b-1 of {@link CloseFutures}. **/
+                if (transferring && closeFutures.notWaitDisconnectClients == null) {
+                    inProgressTransferCloseTask = closeFutures.notDisconnectClients;
+                } else {
+                    if (closeWithoutWaitingClientDisconnect) {
+                        return closeFutures.notWaitDisconnectClients;
+                    }
+                    return closeFutures.waitDisconnectClients;
+                }
+            }
+            fenceTopicToCloseOrDelete();
+            if (!disconnectClients) {
+                this.closeFutures = new CloseFutures(new CompletableFuture(), null, null);
             } else {
-                log.warn("[{}] Topic is already being closed or deleted", topic);
-                closeFuture.completeExceptionally(new TopicFencedException("Topic is already fenced"));
-                return closeFuture;
```

Comment:
> Are we ok to change this behavior? Previously, we made an exception here.

Yes, it is OK.

- When it is called by `pulsar-admin topics unload`, return the result of the previous unloading, which is OK.
- It will not throw an error when it is called by `unload bundle` at the original behavior, because the argument `closeWithoutWaitingClientDisconnect` is `true`